pyo3-ffi/src/lib.rs

```diff
@@ -0,0 +1,457 @@
+mod complexobject;
+#[cfg(all(Py_3_8, not(Py_LIMITED_API)))]
+mod context; // It's actually 3.7.1, but no cfg for patches.
+#[cfg(not(Py_LIMITED_API))]
+pub(crate) mod datetime;
+mod descrobject;
+mod dictobject;
+// skipped dynamic_annotations.h
+mod enumobject;
+// skipped errcode.h
+mod eval;
+// skipped exports.h
+mod fileobject;
+mod fileutils;
+mod floatobject;
+// skipped empty frameobject.h
+#[cfg(not(Py_LIMITED_API))]
+pub(crate) mod funcobject;
+// skipped genericaliasobject.h
+#[cfg(not(Py_LIMITED_API))]
+mod genobject;
+mod import;
+// skipped interpreteridobject.h
+mod intrcheck;
+mod iterobject;
+mod listobject;
+// skipped longintrepr.h
+mod longobject;
+pub(crate) mod marshal;
+mod memoryobject;
+mod methodobject;
+mod modsupport;
+mod moduleobject;
+// skipped namespaceobject.h
+mod object;
+mod objimpl;
+// skipped odictobject.h
+// skipped opcode.h
+// skipped osdefs.h
+mod osmodule;
+// skipped parser_interface.h
+// skipped patchlevel.h
+// skipped picklebufobject.h
+// skipped pyctype.h
+// skipped py_curses.h
+#[cfg(not(Py_LIMITED_API))]
+mod pyarena;
+mod pycapsule;
+// skipped pydecimal.h
+// skipped pydtrace.h
+mod pyerrors;
+// skipped pyexpat.h
+// skipped pyfpe.h
+mod pyframe;
+mod pyhash;
+mod pylifecycle;
+// skipped pymacconfig.h
+// skipped pymacro.h
+// skipped pymath.h
+mod pymem;
+mod pyport;
+mod pystate;
+mod pythonrun;
+// skipped pystrhex.h
+// skipped pystrcmp.h
+mod pystrtod;
+// skipped pythread.h
+// skipped pytime.h
+mod rangeobject;
+mod setobject;
+mod sliceobject;
+mod structseq;
+mod sysmodule;
+mod traceback;
+// skipped tracemalloc.h
+mod tupleobject;
+mod typeslots;
+mod unicodeobject;
+mod warnings;
+mod weakrefobject;
+
+// Additional headers that are not exported by Python.h
+pub mod structmember;
+
+// "Limited API" definitions matching Python's `include/cpython` directory.
+#[cfg(not(Py_LIMITED_API))]
+mod cpython;
+
+#[cfg(not(Py_LIMITED_API))]
+pub use self::cpython::*;
+
+/// This cell type allows shared mutability for python types.
+/// It is intended to hold raw pointers to the python heap and therefore only accepts copy types to
+/// reduce the changes of accidental undefined behaviour.
+/// Thread safety is ensured by the python GIL.
+/// Therefore accessing the contents of this cell in any form while the GIL is not held is UB
+#[repr(transparent)]
+pub struct UnsafeGILCell<T>(Cell<T>);
+
+impl<T> UnsafeGILCell<T> {
+    pub const fn new(val: T) -> UnsafeGILCell<T> {
+        UnsafeGILCell(Cell::new(val))
+    }
+
+    /// Sets the contained value.
+    ///
+    /// # Safety
+    /// The caller must assure that the GIL is held
+    pub unsafe fn set(&self, val: T) {
+        self.0.set(val)
+    }
+}
+
+impl<T: Copy> UnsafeGILCell<T> {
+    /// Returns a copy of the contained value.
+    ///
+    /// # Safety
+    /// The caller must assure that the GIL is held
+    pub unsafe fn get(&self) -> T {
+        self.0.get()
+    }
+}
+
+unsafe impl<T> Sync for UnsafeGILCell<T> {}
```

Comment:
Shouldn't these still include bounds on `T`? Otherwhise while the contract of `UnsafeGILCell` would ensure that updating the shared value is synchronized via the GIL, this would still allow to transfer a `T: !Send` from one thread to another.

Therefore, I think this needs at least
```rust
 unsafe impl<T: Send> Sync for UnsafeGILCell<T> {}
 unsafe impl<T: Send> Send for UnsafeGILCell<T> {}
```
similar to std's [`Mutex`](https://doc.rust-lang.org/stable/std/sync/struct.Mutex.html#impl-Sync).

Reply:
Good point. Trouble is we are storing a raw pointer in the cell... which is not Send (and this is exactly the usecase I have in mind for downstream crates aswell). This is technically still sound because set and get are unsafe anyway so maybe documenting the extra bound would be sufficent?

Reply:
Is the current generality actually required for these use cases? So maybe
```rust
pub struct UnsafeGILCell<T>(Cell<*const T>);
```
would be a better definition and would allow the bounds to be formulated in terms of the pointee `T`? (Which would have to be `Sync` because `&T: Send` if `T: Sync` I guess?)

Reply:
I don't really know much about the datetime api - but this looks like we're reinventing `std::sync::atomic::AtomicPtr`. Can we not use that instead?

Reply:
In principle we do. This is only an optimization insofar it avoids the costs of the atomic accesses if we already have the GIL. (I think the accesses have to be acquire/release pairs at least since we want to publish pointers to objects which can impede some compiler optimizations even though they do not require memory barriers on strongly ordered architectures.)

Reply:
(And taking the GIL is almost surely more expensive if we would have to do that just to access the static.)

Reply:
Also the reason I implemented it this way was that my goal with the API was to mirror the python C API as close as possible (which uses a mutable static) in the pyo3_ffi crate. Other python APIs do the same thing (accessing some mutable static) which we also guard against with require ring the GIL. So it seemed more consistent to me to provide such a raw unsafe mechanism in the pyo3_ffi crate aswell. This also makes a lot of sense because all places where the datetime API is called within pyo3 itself a GIL is always present anyways (so no additional locking mechanism is required). In fact I can't really come up with a usecase where one would want to access this API without holding the GIL. 

With regard to your earlier comment: My other usecase for this generality (in downstream crates) was static `PyTypeObject` instances like they are typically found in python extensions written in C. However I need a pointer to those instances anyway (which would require an UnsafeCell and at that point very few advantages of just a plain static mut) therefore I will follow your suggestion to change this to a 
```
pub struct UnsafeGILCell<T>(Cell<*const T>);

Reply:
After the feedback below I decided not to expose this type after all (see comment below)